ldap/servers/slapd/back-ldbm/cache.c

```diff
@@ -1652,7 +1652,7 @@ cache_lock_entry(struct cache *cache, struct backentry *e)
         /* make sure only one thread does this */
         PR_Lock(cache->c_emutexalloc_mutex);
         if (!e->ep_mutexp) {
-            e->ep_mutexp = PR_NewMonitor();
+            e->ep_mutexp = slapi_pthread_mutex_alloc(PTHREAD_MUTEX_RECURSIVE);
```

Comment:
Recursive mutexes are really risky and generally point to a bigger architectural flaw - and can lead more easily to deadlocks if not handled correctly. 

Especially in the entry cache, which is very high contention, this may actually end up also causing more stalls.

So I'm very skeptical of making this recursive without at least exploring other non-recursive solutions.

Reply:
It's an interesting point...
I've done this PR because I wanted to bring consistency between this change: https://github.com/389ds/389-ds-base/pull/4569
And this function (where we use both entry cache mutex and entry mutex):
https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/cache.c#L1647

This change is an attempt to fix this issue (and improve performance): https://github.com/389ds/389-ds-base/issues/5362 (which is not 100% reproducible, BTW, - I've tried a lot, and with high entry cache pressure - no luck)

Which happens (if I understand correctly) because we are not able to lock the loaded entry in time. And **LDBM_CACHE_RETRY_COUNT if** blocks us because we tried 1000 times:
https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/findentry.c#L142

```
[06/Jul/2022:19:05:34.236431632 +0000] - ERR - find_entry_internal_dn - Retry count exceeded (cn=services,cn=accounts,dc=ipa,dc=test)
```

So I thought it may happen because NSPR monitor and thread recursive mutex have very different dynamics.

But... Could it be that the initial change caused it - https://github.com/389ds/389-ds-base/pull/4569 - because the recursive mutex has this ownership thing when the thread that grabs the mutex must be the same thread that releases the mutex? So I wonder if we may have a rare deadlock because of the recursive mutex nature.

I also thought about using atomics with AQUIRE/RELEASE so we have load/lock here - https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/findentry.c#L103 - happen atomically. But... It may cause bigger performance problems.

Reply:
> It's an interesting point... I've done this PR because I wanted to bring consistency between this change: #4569 And this function (where we use both entry cache mutex and entry mutex): https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/cache.c#L1647
> 
> This change is an attempt to fix this issue (and improve performance): #5362 (which is not 100% reproducible, BTW, - I've tried a lot, and with high entry cache pressure - no luck)
> 
> Which happens (if I understand correctly) because we are not able to lock the loaded entry in time. And **LDBM_CACHE_RETRY_COUNT if** blocks us because we tried 1000 times: https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/findentry.c#L142
> 
> ```
> [06/Jul/2022:19:05:34.236431632 +0000] - ERR - find_entry_internal_dn - Retry count exceeded (cn=services,cn=accounts,dc=ipa,dc=test)
> ```
> 
> So I thought it may happen because NSPR monitor and thread recursive mutex have very different dynamics.

NSPR montior and pthread locks do have different dynamics, but being recursive may not help .... 

> 
> But... Could it be that the initial change caused it - #4569 - because the recursive mutex has this ownership thing when the thread that grabs the mutex must be the same thread that releases the mutex? So I wonder if we may have a rare deadlock because of the recursive mutex nature.

Yes, it's possible to deadlock like this. 

> 
> I also thought about using atomics with AQUIRE/RELEASE so we have load/lock here - https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/findentry.c#L103 - happen atomically. But... It may cause bigger performance problems.

Nooooppppppeeeeeee don't do that, almost every use of atomics from older code in the code base is wrong. It won't just be performance issues, but stability ones.

Reply:
A few points:
  - Do not be worried about the "thread that grabs the mutex must be the same thread that releases the mutex" condition:
    In fact this condition must be true with pthread lock whether is mutex is recursive or not.
     ( The "Unlock When Not Owner" column in pthread_mutex_lock man page is quite clear! )
    And as far as I know ns-slapd and the plugin respects this condition
  - I do not share @firstyear concerns about recursive mutex (as far as the mutex implementation is correct)
     there is not much differences between a regular mutex and a recursive one 
     (IMHO, the main concern is that deadlocks may be a bit more painful to debug)
     but personally I am much more warry of read write locks than of recursive mutexes.
  - Like @firstyear, I do not think that atomic operation is the right solution to your problem:
     using atomic operation usually means having a retry loop and IMHO you risk to have the same issue 
     than the one you are trying to solve.
  - I am a bit surprise that you speak about deadlock as possible cause because that does not match the 
     customer symptoms (Is it what you got after replacing NSPR per pthread mutex ? )

Reply:
So I've tried to reproduce the entry cache issue in different ways...

1. Modify a user entry's description attribute with a lot of modify operations. At the same time as there was a tombstone trimming set to a very small value.
2. Do add/delete operations in a loop and then try to search the same entry (still while tombstone trimming is active). No error.
3. I've run the FreeIPA test a few times, but still no luck.

We saw the error on FreeIPA PR CI tests with promoting client to replica, and it may indicate that the issue is not about heavy load but because something is wrong in the entry cache mutex interactions. And we know for sure that the issue doesn't happen if we go back to NSPR Monitor (so away from pthread recursive mutex).

We may revert all of the pthread recursive changes back to NSPR Monitor. We improve the stability but lose the performance gain (we had 8% search improvement after moving to pthread recursive).
I need to think more...

Reply:
> * I do not share @Firstyear concerns about recursive mutex (as far as the mutex implementation is correct)
>   there is not much differences between a regular mutex and a recursive one
>   (IMHO, the main concern is that deadlocks may be a bit more painful to debug)
>   but personally I am much more warry of read write locks than of recursive mutexes.

Given that the PR_Lock is not recursive, why do we need to add recursion now? 

The recursive counter adds some overhead I thought. I think to start with changing to pthread mutex is the right start, then add recursion *later* if it's actually required. 

> * Like @Firstyear, I do not think that atomic operation is the right solution to your problem:
>   using atomic operation usually means having a retry loop and IMHO you risk to have the same issue
>   than the one you are trying to solve.

^- this. And it's also infinitely harder to get write on weakly ordered CPU arches. 

> * I am a bit surprise that you speak about deadlock as possible cause because that does not match the
>   customer symptoms (Is it what you got after replacing NSPR per pthread mutex ? )

Reply:
@Firstyear but we've changed PRMonitor, not PRLock... And PRMonitor works with counts, the same as pthread mutex recursive. Or do I miss something?

Regarding the atomic part - sure, makes 100% sense. :) I agree.

Reply:
You're right, that's my mistake. I thought it was PR_lock.